### PR TITLE
Implement promotion method

### PR DESCRIPTION
**Pull request by eray-felek-sonarsource**

### CODE REVIEW

File: src/main/java/org/sonarlint/intellij/SonarLintIntelliJClient.kt [hunk 1]
```diff
@@ -726,7 +727,12 @@ object SonarLintIntelliJClient : SonarLintRpcClientDelegate {
     }
 
     override fun promoteExtraEnabledLanguagesInConnectedMode(configurationScopeId: String, languagesToPromote: Set<Language>) {
-        // TODO: Move 'raw' analysis to SLCORE
+        if (languagesToPromote.isEmpty()) return
+
+        val project = BackendService.findModule(configurationScopeId)?.project
+            ?: BackendService.findProject(configurationScopeId) ?: return
+
+        project.getMessageBus().syncPublisher<PromotionListener>(PromotionListener.TOPIC).extraLanguagePromotionRaised(languagesToPromote)
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/SonarLintIntelliJClient.kt:735]: I think you can just directly call the methods to show the notification here; you don't need to create a listener?

File: src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt [hunk 2]
```diff
@@ -188,7 +200,21 @@ class PromotionProvider(private val project: Project) {
...
+            languages.find {
+                it.name == language.name
+            }?.let { extraLanguagesToPromote.add(org.sonarsource.sonarlint.core.client.utils.Language.fromDto(it).label) }
+        }
+
+        return extraLanguagesToPromote
+    }
+
     private fun showPromotion(notifications: SonarLintProjectNotifications, content: String) {
+        PropertiesComponent.getInstance().setValue(LAST_PROMOTION_NOTIFICATION_DATE, Instant.now().toEpochMilli().toString())
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt:216]: Is this change needed? It's all done after in the `if()`, I think there's an issue

**Review by nquinquenel — COMMENTED** [review]
Looks good but I think it can be simplified

File: src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt (hunk 2 above)
Comment (eray-felek-sonarsource) [reply src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt]: It was leftover code while testing forgot to delete

File: src/main/java/org/sonarlint/intellij/SonarLintIntelliJClient.kt (hunk 1 above)
Comment (eray-felek-sonarsource) [reply src/main/java/org/sonarlint/intellij/SonarLintIntelliJClient.kt:735]: PromotionProvider is a project-level service. I had initially implemented to use as methods and converted to this design to have separation of concerns. Calling the method would actually require to change the design and I think it would not be a good design choice.

Comment (nquinquenel) [reply src/main/java/org/sonarlint/intellij/SonarLintIntelliJClient.kt:735]: `PromotionProvider` is currently based on listener because it fits what it needs to do: listen to a file opened and when an analysis is started.

Here we are already listening to a kind of event: we receive a request to `promoteExtraEnabledLanguagesInConnectedMode` when needed from the backend, it does not make much sense to add a listener on top of this.

It's also not an issue if it's a project-level service, we have access to the project so you can directly call getService(project, PromotionProvider::class.java)

File: src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt [hunk 3]
```diff
@@ -188,6 +200,18 @@ class PromotionProvider(private val project: Project) {
...
+    private fun findExtraLanguages(languagesToPromote: Set<Language>, languages: Set<Language>): List<String> {
+        val extraLanguagesToPromote = ArrayList<String>()
+
+        languagesToPromote.forEach { language ->
+            languages.find {
+                it.name == language.name
+            }?.let { extraLanguagesToPromote.add(org.sonarsource.sonarlint.core.client.utils.Language.fromDto(it).label) }
+        }
+
+        return extraLanguagesToPromote
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt:212]:
```suggestion
        return languagesToPromote.mapNotNull { language ->
            languages.find {
                it.name == language.name
            }?.let { org.sonarsource.sonarlint.core.client.utils.Language.fromDto(it).label }
        }.toList()
```

File: src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt [hunk 4]
```diff
@@ -172,12 +182,14 @@ class PromotionProvider(private val project: Project) {
...
-        val language = findLanguage(extension, extraEnabledLanguagesInConnectedMode)
+    private fun processExtraLanguagePromotion(notifications: SonarLintProjectNotifications, languagesToPromote: Set<Language>) {
+        val languages = findExtraLanguages(languagesToPromote, extraEnabledLanguagesInConnectedMode)
 
-        if (language != null) {
-            showPromotion(notifications, "Enable " + language.label + " analysis by connecting your project")
-        }
+        if (languages.isEmpty()) return
+
+        val languagesText = languages.joinToString(", ")
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/promotion/PromotionProvider.kt:190]: Maybe to stay consistent with SLE, we can do the following:
```suggestion
        val languagesText = languages.joinToString(" / ")
```

File: src/main/java/org/sonarlint/intellij/SonarLintIntelliJClient.kt (hunk 1 above)
Comment (eray-felek-sonarsource) [reply src/main/java/org/sonarlint/intellij/SonarLintIntelliJClient.kt]: Makes sense, changed to use method

**Review by nquinquenel — APPROVED** [review]
LGTM!
